modules/generator/registry/native_histogram.go

```diff
@@ -301,11 +319,64 @@ func (h *nativeHistogram) nativeHistograms(appender storage.Appender, lbls label
 	}
 	hist.NegativeBuckets = s.histogram.NegativeDelta
 
-	_, err = appender.AppendHistogram(0, lbls, timeMs, &hist, nil)
+	// Append the native histogram
+	ref, err := appender.AppendHistogram(0, lbls, timeMs, &hist, nil)
 	if err != nil {
 		return err
 	}
 
+	// NOTE: two exemplar formats are used:
+	// Native exemplars use the histogram.Exemplars slice, which is the native format.
+	// Bucket exemplars use the bucket.Exemplar field, which is the classic format.
+	//
+	// Use native exemplars when available, falling back to bucket exemplars
+
+	nativeExemplarCount := 0
+	for _, ex := range s.histogram.Exemplars {
+		if ex != nil && len(ex.Label) > 0 {
+			nativeExemplarCount++
+			_, err = appender.AppendExemplar(ref, lbls, exemplar.Exemplar{
+				Labels: convertLabelPairToLabels(ex.GetLabel()),
+				Value:  ex.GetValue(),
+				Ts:     timeMs,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// NOTE: We clear the native exemplar slice to prevent accumulation, but the
+	// client_golang package handles the expiration of exemplars internally, and
+	// we don't have control over clearing the native histogram exemplars in the
+	// same way we do for the class histogram exemplars.
+	if nativeExemplarCount > 0 {
```

Comment:
It might be better to have this check `len(s.histogram.Exemplars) > 0` instead, because `nativeExemplarCount` is only incremented if Label is present. Not sure if that matters, but this way it would always reset a non-empty slice.

Reply:
Good call, this is cleaner.